### PR TITLE
Assignment7+8 linq @avanchu253

### PR DESCRIPTION
**Pull request by randypipkins**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -46,48 +46,48 @@ Chapter 14: Events
 
 **Throughout, consider using the `System.Linq.Enumerable` methods `Zip`, `Count`, `Sort` and `Contains` methods for testing collections.**. (Preferably avoid using `Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert` although that might be easier, to get a firmer grasp on additional LINQ API.)
 
-1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ❌✔
+1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔
 
-   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌✔
-   - Using LINQ, skip the first row in the `People.csv`. ❌✔
-   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ❌✔
+   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
```
Comment (alowe4) [inline README.md:51]: this was not done, you need to right click on the people.csv file and change the properties to Copy if newer, changes should be reflected in your Assignment.csproj file

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            Assert.AreEqual<string>("1,Priscilla,Jenyns,pjenyns0@state.gov,7884 Corry Way,Helena,MT,70577", rows[0]);
+        }
+
+        [TestMethod]
+        public void SortedListOfStatesIsUnique()
+        {
+            SampleData sample = new();
+
+            IEnumerable<string> states = sample.GetUniqueSortedListOfStatesGivenCsvRows();
+
```
Comment (alowe4) [inline Assignment/Assignment.Tests/SampleDataTests.cs:28]: this test uses LINQ to verify that the GetUniqueSortedListOfStatesGivenCsvRows method works, you also need to use a hard coded list of addresses to verify that the GetUniqueSortedListOfStatesGivenCsvRows method works

**Review by alowe4 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
        Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌
        Using LINQ, skip the first row in the People.csv. ✔
        Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

  Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
        Use ISampleData.CsvRows for your data source. ✔
        Don't forget the list should be unique. ✔
        Sort the list alphabetically. ✔
        Include a test that leverages a hardcoded list of addresses. ❌
        Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

   Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
        Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
        Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
        Given the array, consider using string.Join to combine the list into a single string. ✔

   Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
        Use ISampleData.CsvRows as the source of the data. ✔
        Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
        Be sure that Person.Address is also populated. ✔
        Adding null validation to all the Person and Address properties is optional.
        Consider using ISampleData.CsvRows in your test to verify your results. ❌

   Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
        Use ISampleData.People for your data source. ✔

   Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
        Use the people parameter from ISampleData.People property for your data source. ✔
        At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
        Don't forget the list should be unique. ✔
        It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.❌

   Given the implementation of Node in Assignment5

   Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
    Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.✔

Extra Credit

   Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals

   Place all shared project properties into a Directory.Build.Props file.✔
   Place all shared project items into a Directory.Build.targets file.✔
    Ensure nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

**Comment by Clockknight** [comment]
I will review this.

**Comment by Clockknight** [comment]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

Implement IEnumerable GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ❌
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔

Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. 

Implement ISampleDate.FilterByEmailAddress(Predicate filter) to return a list of names where the email address matches the filter. ✔
Use ISampleData.People for your data source. ✔

Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable people) to return a string that contains a unique, comma-separated list of states. ✔
Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.❌

Given the implementation of Node in Assignment5

Implement IEnumerable to return all the items in the "circle" of items. ✔
Add an IEnumberable ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.✔

Fundamentals

Place all shared project properties into a Directory.Build.Props file.✔
Place all shared project items into a Directory.Build.targets file.✔
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔


Seems good for the most part! From what I can tell, you didn't attempt the extra credit, but seemed to pass the rest of it!

File: Assignment/Assignment.Tests/NodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Assignment.Tests
+{
+    [Microsoft.VisualStudio.TestTools.UnitTesting.TestClass]
+    public class NodeTests
+    {
+        [TestMethod]
+        public void ReturnsChildItemsGivenMax()
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/NodeTests.cs:15]: You could probably use a few more tests here to test different amounts. You may want to also test that you are circling back correctly and grabbing the correct nodes.

**Review by mmwoodfo — COMMENTED** [review]
This should be set to copy if newer
image

**Pull request closed without merging**